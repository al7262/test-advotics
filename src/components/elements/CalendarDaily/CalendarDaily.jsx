/* eslint-disable react/no-array-index-key */
/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useEffect, useState } from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import styles from './index.css';

export default function CalendarDaily(props) {
  const {
    date, disabledAfter, disabledBefore, name, onSelect,
  } = props;
  const objDay = date ? moment(date) : moment();
  const objAfter = moment(disabledAfter);
  const objBefore = moment(disabledBefore);
  const toYear = objDay.year();
  const toMonth = objDay.month();
  const [year, setYear] = useState(toYear);
  const [month, setMonth] = useState(toMonth);
  const months = moment.months();
  const weekdays = moment.weekdaysMin();
  const days = moment().year(year).month(month).daysInMonth();
  const firstDay = moment().year(year).month(month).date(1)
    .day();
  const firstDate = moment(firstDay).add(-1, 'days').date();
  const lastDay = 7 - moment().year(year).month(month + 1).date(1)
    .day();

  const onClickArrow = (next) => () => {
    if (next && month === 11) {
      setMonth(0);
      setYear(year + 1);
    } else if (next) {
      setMonth(month + 1);
    } else if (month === 0) {
      setMonth(11);
      setYear(year - 1);
    } else {
      setMonth(month - 1);
    }
  };

  useEffect(() => {
    setMonth(toMonth);
    setYear(toYear);
  }, [date]);

  const onClickDate = (i) => () => onSelect(moment().year(year).month(month).date(i + 1)
    .format('YYYY-MM-DD'));

  return (
    <div className={styles.root}>
      <label>{name}</label>
      <button type="button" onClick={onClickArrow(false)}>
        <span className={styles['arrow-left']} />
      </button>
      <b>
        {`${months[month]} ${year}`}
      </b>
      <button type="button" onClick={onClickArrow(true)}>
        <span className={styles['arrow-right']} />
      </button>
      {weekdays.map((i) => <p key={i}>{i}</p>)}
      <span className={styles.horizontal} />
      {[...Array.from({ length: firstDay }).keys()]
        .map((i) => <span key={`prev-${i}`}>{firstDate + 1 - (firstDay - i)}</span>)}
      {[...Array.from({ length: days }).keys()].map((i, idx) => {
        const objDayLoop = moment().year(year).month(month).date(i + 1);
        const selected = objDay.isSame(objDayLoop, 'day');
        const isDisabledAfter = disabledAfter && objAfter.isBefore(objDayLoop, 'day');
        const isDisabledBefore = disabledBefore && objBefore.isAfter(objDayLoop, 'day');
        if (isDisabledAfter) {
          return <span key={idx}>{i + 1}</span>;
        }
        if (isDisabledBefore) {
          return <span key={idx}>{i + 1}</span>;
        }
        return (
          <button
            type="button"
            disabled={date && selected}
            key={idx}
            onClick={onClickDate(i)}
          >
            {i + 1}
          </button>
        );
      })}
      {[...Array.from({ length: lastDay }).keys()].map((i) => (
        lastDay < 7 && <span key={`next-${i}`}>{lastDay + 1 - (lastDay - i)}</span>
      ))}
    </div>
  );
}

CalendarDaily.defaultProps = {
  date: '',
  disabledAfter: '',
  disabledBefore: '',
  name: '',
  onSelect: () => { },
};

CalendarDaily.propTypes = {
  date: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  disabledAfter: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  disabledBefore: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  name: PropTypes.string,
  onSelect: PropTypes.func,
};
