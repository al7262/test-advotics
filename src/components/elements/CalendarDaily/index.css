.root {
  align-items: center;
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  justify-items: center;
}

.root > label {
  grid-column: 1 / span 7;
  font-size: 1rem;
  font-weight: normal;
  text-transform: uppercase;
}

.root > b {
  grid-column: 2 / span 5;
  font-size: .875rem;
}

.root > p {
  font-size: 0.75rem;
  font-weight: bold;
  margin: .5rem;
  color: var(--black-color-3)
}

.root > span {
  color: var(--black-color-4);
  font-size: 0.75rem;
  line-height: 1rem;
  width: 1rem;
  text-align: center;
  margin: .375rem;
}

.root > .horizontal {
  grid-column: 1 / span 7;
  width: 100%;
  height: .0625rem;
  background-color: var(--black-color-4);
  margin: 0;
}

.root > button {
  background-color: transparent;
  border: none;
  cursor: pointer;
  font-size: 0.75rem;
  height: 1.25rem;
  margin: .25rem;
  width: 1.25rem;
  position: relative;
}

.root > button:disabled {
  color: var(--primary-color);
  cursor: default;
  font-weight: 700;
}

.root > button:disabled::after {
  content: "";
  background-color: var(--primary-color);
  border: .0625rem solid var(--primary-color-dark);
  border-radius: 50%;
  width: 1.5rem;
  height: 1.5rem;
  position: absolute;
  top: -.125rem;
  left: -.125rem;
  opacity: .25;
}

.root > button:hover:not(:disabled) {
  color: var(--primary-color);
}

.root > button:first-of-type {
  grid-column: 1;
  justify-self: self-start;
  display: flex;
  padding-top: .375rem;
  padding-left: .5rem;
}

.root > button:nth-of-type(2) {
  grid-column: 7;
  justify-self: self-end;
  display: flex;
  padding-top: .375rem;
  padding-left: .25rem;
}

.root > button:first-of-type > img,
.root > button:nth-of-type(2) > img {
  filter: brightness(0.3);
}

.arrow-left {
  border: solid var(--black-color-2);
  border-width: 0 .125rem .125rem 0;
  display: inline-block;
  padding: .2rem;
  transform: rotate(135deg);
}

.arrow-right {
  border: solid var(--black-color-2);
  border-width: 0 .125rem .125rem 0;
  display: inline-block;
  padding: .2rem;
  transform: rotate(-45deg);
}
