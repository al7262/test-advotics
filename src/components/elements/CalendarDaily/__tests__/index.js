/* eslint-disable react/jsx-filename-extension */
/* eslint-disable no-undef */
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import CalendarDaily from '../CalendarDaily';

describe('src/components/elements/CalendarDaily', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<CalendarDaily />);
    expect(tree).toMatchSnapshot();
  });
});
