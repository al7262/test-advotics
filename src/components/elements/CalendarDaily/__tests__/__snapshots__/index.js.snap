// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/CalendarDaily render 1`] = `
<div>
  <label>
    
  </label>
  <button
    onClick={[Function]}
    type="button"
  >
    <span />
  </button>
  <b>
    July 2021
  </b>
  <button
    onClick={[Function]}
    type="button"
  >
    <span />
  </button>
  <p>
    Su
  </p>
  <p>
    Mo
  </p>
  <p>
    Tu
  </p>
  <p>
    We
  </p>
  <p>
    Th
  </p>
  <p>
    Fr
  </p>
  <p>
    Sa
  </p>
  <span />
  <span>
    28
  </span>
  <span>
    29
  </span>
  <span>
    30
  </span>
  <span>
    31
  </span>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    6
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    7
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    8
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    9
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    10
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    11
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    12
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    13
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    14
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    15
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    16
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    17
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    18
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    19
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    20
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    21
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    22
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    23
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    24
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    25
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    26
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    27
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    28
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    29
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    30
  </button>
  <button
    disabled=""
    onClick={[Function]}
    type="button"
  >
    31
  </button>
</div>
`;
